# 📚 Libraries & Data Structure

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.simplefilter("ignore")
from sklearn.preprocessing import StandardScaler

In [2]:
loan_export = pd.read_csv('LoanExport.csv', low_memory=False)

In [3]:
loan_export.columns

Index(['CreditScore', 'FirstPaymentDate', 'FirstTimeHomebuyer', 'MaturityDate',
       'MSA', 'MIP', 'Units', 'Occupancy', 'OCLTV', 'DTI', 'OrigUPB', 'LTV',
       'OrigInterestRate', 'Channel', 'PPM', 'ProductType', 'PropertyState',
       'PropertyType', 'PostalCode', 'LoanSeqNum', 'LoanPurpose',
       'OrigLoanTerm', 'NumBorrowers', 'SellerName', 'ServicerName',
       'EverDelinquent', 'MonthsDelinquent', 'MonthsInRepayment'],
      dtype='object')

In [4]:
loan_export.sample(7)

CreditScore  FirstPaymentDate FirstTimeHomebuyer  MaturityDate    MSA  \
20228           630            199903                  X        202902  X       
108430          697            199903                  X        202902  26900   
206578          747            199904                  X        202903  35084   
115729          701            199904                  N        202903  35840   
211333          749            199904                  X        202903  40060   
13876           620            199905                  N        202904  X       
247127          766            199904                  X        202903  48424   

        MIP  Units Occupancy  OCLTV  DTI  ...  PostalCode    LoanSeqNum  \
20228     0      1         O     77   28  ...       35000  F199Q1307273   
108430    0      1         O     74    0  ...       46000  F199Q1379748   
206578    0      1         O     18   50  ...        7900  F199Q1034701   
115729   25      1         O     90   34  ...       34200  F199Q1398026   
211333    0      1         O     68   14  ...       23800  F199Q1082597   
13876    12      1         O     85   36  ...       49300  F199Q1388237   
247127    0      1         O     80   42  ...       33400  F199Q1397009   

        LoanPurpose OrigLoanTerm NumBorrowers SellerName  \
20228             N          360            1         Ot   
108430            P          360            1         FI   
206578            C          360            1         PN   
115729            P          360            2         Ot   
211333            C          360            2         OL   
13876             N          360            2         GM   
247127            C          360            1         Ot   

                ServicerName EverDelinquent MonthsDelinquent MonthsInRepayment  
20228   WASHINGTONMUTUALBANK              0                0                50  
108430  Other servicers                   1                4               141  
206578   JPMORGANCHASEBANKNA              1                2               159  
115729   SUNTRUSTMORTGAGEINC              1                9               165  
211333  Other servicers                   0                0                 8  
13876   GMACMTGECORP                      1                4                54  
247127   SUNTRUSTMORTGAGEINC              0                0                41  

[7 rows x 28 columns]

In [5]:
loan_export.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 291451 entries, 0 to 291450
Data columns (total 28 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   CreditScore         291451 non-null  int64  
 1   FirstPaymentDate    291451 non-null  int64  
 2   FirstTimeHomebuyer  291451 non-null  object 
 3   MaturityDate        291451 non-null  int64  
 4   MSA                 291451 non-null  object 
 5   MIP                 291451 non-null  int64  
 6   Units               291451 non-null  int64  
 7   Occupancy           291451 non-null  object 
 8   OCLTV               291451 non-null  int64  
 9   DTI                 291451 non-null  int64  
 10  OrigUPB             291451 non-null  int64  
 11  LTV                 291451 non-null  int64  
 12  OrigInterestRate    291451 non-null  float64
 13  Channel             291451 non-null  object 
 14  PPM                 291451 non-null  object 
 15  ProductType         291451 non-nul

In [6]:
numerical_attribs, categorical_attribs = loan_export.describe(exclude=[object]), loan_export.describe(exclude=[np.number])

In [7]:
categorical, numerical = loan_export[categorical_attribs.columns], loan_export[numerical_attribs.columns]

In [8]:
numerical.describe()

CreditScore  FirstPaymentDate   MaturityDate            MIP  \
count  291451.000000     291451.000000  291451.000000  291451.000000   
mean      708.936991     199904.760553  202902.367043       9.216685   
std        68.408791         13.821228      16.090508      12.648273   
min         0.000000     199901.000000  202402.000000       0.000000   
25%       675.000000     199903.000000  202902.000000       0.000000   
50%       717.000000     199904.000000  202903.000000       0.000000   
75%       753.000000     199905.000000  202904.000000      25.000000   
max       849.000000     201303.000000  204302.000000      55.000000   

               Units          OCLTV            DTI        OrigUPB  \
count  291451.000000  291451.000000  291451.000000  291451.000000   
mean        1.026022      76.960529      30.033453  124940.387921   
std         0.202986      15.236154      13.203974   53657.440624   
min         0.000000       0.000000       0.000000    8000.000000   
25%         1.000000      70.000000      22.000000   83000.000000   
50%         1.000000      80.000000      31.000000  117000.000000   
75%         1.000000      90.000000      39.000000  160000.000000   
max         4.000000     110.000000      65.000000  497000.000000   

                 LTV  OrigInterestRate   OrigLoanTerm  EverDelinquent  \
count  291451.000000     291451.000000  291451.000000   291451.000000   
mean       76.937986          6.926547     359.835375        0.197848   
std        15.225298          0.343541       2.286939        0.398378   
min         0.000000          4.000000     301.000000        0.000000   
25%        70.000000          6.750000     360.000000        0.000000   
50%        80.000000          6.875000     360.000000        0.000000   
75%        90.000000          7.125000     360.000000        0.000000   
max       103.000000         12.350000     361.000000        1.000000   

       MonthsDelinquent  MonthsInRepayment  
count     291451.000000      291451.000000  
mean           1.837729          60.805291  
std            8.809392          46.142599  
min            0.000000           1.000000  
25%            0.000000          34.000000  
50%            0.000000          48.000000  
75%            0.000000          63.000000  
max          180.000000         212.000000

In [9]:
categorical.describe()

FirstTimeHomebuyer     MSA Occupancy Channel     PPM ProductType  \
count              291451  291451    291451  291451  291451      291451   
unique                  3     392         3       4       3           1   
top                     N   X             O       T       N       FRM     
freq               184154   39100    276131  167670  282125      291451   

       PropertyState PropertyType PostalCode    LoanSeqNum LoanPurpose  \
count         291451       291451     291451        291451      291451   
unique            53            7        892        291451           3   
top               CA           SF      94500  F199Q1268030           P   
freq           43327       244923       3776             1      123404   

       NumBorrowers SellerName          ServicerName  
count        291451     266457                291451  
unique            3         20                    20  
top               2         Ot  Other servicers       
freq         187335      76943                 64043

# 🧑‍💻 Preprocessing

### 1. Drop Irrelivvant Columns

* Drop irrelevant columns:
    - The dataset used contains information that is unavailable at the time of loan application. We will drop these columns before starting with our analysis. The columns we will drop are: `FirstPaymentDate` `MaturityDate` `MIP` `OrigUPB` `OrigInterestRate` `PPM` 
    - The column that i decide to drop it:
        - `LTV:` the correlation between it and OCLTV so must drop one of them
        - `SellerName:` there is 24994 NAN values from 291451
        - `OrigLoanTerm` there are positive and negative correlations between it and another two column
        - `PropertyState`as this information is encoded in the MSA column.
        - `LoanSeqNum` unique id assigned to each loan. As it provides no information we will drop this column.
        - `ServicerName` dependent loan activity and since this information is not available at the time of loan request we will drop these columns.
        - `ProductType` this column has the same value for all rows, it provides no useful information to us.

In [10]:
loan_export_copy = loan_export.copy()

In [11]:
def drop_columns(df):
    coulmns = [
        'LTV', 'SellerName', 'OrigLoanTerm', 'FirstPaymentDate',
        'MaturityDate', 'MIP', 'OrigUPB', 'OrigInterestRate',
        'PPM','PropertyState', 'LoanSeqNum', 'ServicerName',
        'ProductType',
    ]
    df.drop(coulmns, axis=1, inplace=True)
    
    return df

In [12]:
loan_export_copy = drop_columns(loan_export_copy)

### 2. Missing Values

In [13]:
categorical.FirstTimeHomebuyer.value_counts()

N    184154
X     78015
Y     29282
Name: FirstTimeHomebuyer, dtype: int64

In [14]:
# replace X with NA
a = categorical['FirstTimeHomebuyer'].values
categorical['FirstTimeHomebuyer'] = np.where(a == 'X', np.NaN, a)

In [15]:
categorical.FirstTimeHomebuyer.value_counts()

N    184154
Y     29282
Name: FirstTimeHomebuyer, dtype: int64

In [16]:
categorical.FirstTimeHomebuyer.isna().sum()

78015

In [17]:
categorical.MSA.value_counts()

X        39100
31084     9338
16974     8771
12060     6985
47644     6673
         ...  
21940        3
49500        2
25020        1
10380        1
32420        1
Name: MSA, Length: 392, dtype: int64

In [18]:
categorical['MSA'] = categorical['MSA'].replace('X    ', 'X')

In [19]:
# replace X with NA
a = categorical['MSA'].values
categorical['MSA'] = np.where(a == 'X', np.NaN, a)

In [20]:
categorical['MSA'].value_counts()

31084    9338
16974    8771
12060    6985
47644    6673
38060    6201
         ... 
21940       3
49500       2
25020       1
10380       1
32420       1
Name: MSA, Length: 391, dtype: int64

In [21]:
def replace_X_with_NAN(df):
    columns = df.select_dtypes(include=['object']).columns
    for column in columns:
        df.loc[df[column].str.strip() == 'X', column] = np.nan
    return df

In [22]:
categorical = replace_X_with_NAN(categorical)

In [23]:
categorical['MSA'].value_counts()

31084    9338
16974    8771
12060    6985
47644    6673
38060    6201
         ... 
21940       3
49500       2
25020       1
10380       1
32420       1
Name: MSA, Length: 391, dtype: int64

In [24]:
def print_value_counts(data_frame):
    for column in data_frame.columns:
        print(f"Column: {column}")
        print(data_frame[column].value_counts())
        print("#" * 50)

In [25]:
print_value_counts(categorical)

Column: FirstTimeHomebuyer
N    184154
Y     29282
Name: FirstTimeHomebuyer, dtype: int64
##################################################
Column: MSA
31084    9338
16974    8771
12060    6985
47644    6673
38060    6201
         ... 
21940       3
49500       2
25020       1
10380       1
32420       1
Name: MSA, Length: 391, dtype: int64
##################################################
Column: Occupancy
O    276131
I      8734
S      6586
Name: Occupancy, dtype: int64
##################################################
Column: Channel
T    167670
R    123219
C       407
B       155
Name: Channel, dtype: int64
##################################################
Column: PPM
N    282125
Y      3921
Name: PPM, dtype: int64
##################################################
Column: ProductType
FRM      291451
Name: ProductType, dtype: int64
##################################################
Column: PropertyState
CA    43327
FL    19133
MI    16285
IL    13621
TX    13092
OH    11545
CO 

In [26]:
categorical.isna().sum()

FirstTimeHomebuyer    78015
MSA                   39100
Occupancy                 0
Channel                   0
PPM                    5405
ProductType               0
PropertyState             0
PropertyType             22
PostalCode                6
LoanSeqNum                0
LoanPurpose               0
NumBorrowers            339
SellerName            24994
ServicerName              0
dtype: int64

In [27]:
numerical.isna().sum()

CreditScore          0
FirstPaymentDate     0
MaturityDate         0
MIP                  0
Units                0
OCLTV                0
DTI                  0
OrigUPB              0
LTV                  0
OrigInterestRate     0
OrigLoanTerm         0
EverDelinquent       0
MonthsDelinquent     0
MonthsInRepayment    0
dtype: int64

In [28]:
loan_export_copy = replace_X_with_NAN(loan_export_copy)

In [29]:
loan_export_copy.isna().sum()

CreditScore               0
FirstTimeHomebuyer    78015
MSA                   39100
Units                     0
Occupancy                 0
OCLTV                     0
DTI                       0
Channel                   0
PropertyType             22
PostalCode                6
LoanPurpose               0
NumBorrowers            339
EverDelinquent            0
MonthsDelinquent          0
MonthsInRepayment         0
dtype: int64

In [30]:
columns_with_na = ['MSA', 'FirstTimeHomebuyer', 'PropertyType', 'NumBorrowers', 'PostalCode']

In [31]:
print_value_counts(loan_export_copy[columns_with_na])

Column: MSA
31084    9338
16974    8771
12060    6985
47644    6673
38060    6201
         ... 
21940       3
49500       2
25020       1
10380       1
32420       1
Name: MSA, Length: 391, dtype: int64
##################################################
Column: FirstTimeHomebuyer
N    184154
Y     29282
Name: FirstTimeHomebuyer, dtype: int64
##################################################
Column: PropertyType
SF    244923
PU     27506
CO     18100
MH       723
LH       105
CP        72
Name: PropertyType, dtype: int64
##################################################
Column: NumBorrowers
2    187335
1    103777
Name: NumBorrowers, dtype: int64
##################################################
Column: PostalCode
94500    3776
30000    3637
85200    3280
48100    3246
48000    2988
         ... 
34500       1
4700        1
33200       1
84200       1
41300       1
Name: PostalCode, Length: 891, dtype: int64
##################################################


In [32]:
def preprocess_loan_export(data_frame):
    data_frame['FirstTimeHomebuyer'].fillna('N', inplace=True)
    data_frame['PropertyType'].fillna('Not Available', inplace=True)
    data_frame['NumBorrowers'].fillna('Not Available', inplace=True)
    data_frame['MSA'].fillna(0, inplace=True)
    data_frame.dropna(subset=['PostalCode'], inplace=True)
    return data_frame

In [33]:
loan_export_copy = preprocess_loan_export(loan_export_copy)

In [34]:
loan_export_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291445 entries, 0 to 291450
Data columns (total 15 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   CreditScore         291445 non-null  int64 
 1   FirstTimeHomebuyer  291445 non-null  object
 2   MSA                 291445 non-null  object
 3   Units               291445 non-null  int64 
 4   Occupancy           291445 non-null  object
 5   OCLTV               291445 non-null  int64 
 6   DTI                 291445 non-null  int64 
 7   Channel             291445 non-null  object
 8   PropertyType        291445 non-null  object
 9   PostalCode          291445 non-null  object
 10  LoanPurpose         291445 non-null  object
 11  NumBorrowers        291445 non-null  object
 12  EverDelinquent      291445 non-null  int64 
 13  MonthsDelinquent    291445 non-null  int64 
 14  MonthsInRepayment   291445 non-null  int64 
dtypes: int64(7), object(8)
memory usage: 35.6+ MB


In [35]:
loan_export_copy.isna().sum()

CreditScore           0
FirstTimeHomebuyer    0
MSA                   0
Units                 0
Occupancy             0
OCLTV                 0
DTI                   0
Channel               0
PropertyType          0
PostalCode            0
LoanPurpose           0
NumBorrowers          0
EverDelinquent        0
MonthsDelinquent      0
MonthsInRepayment     0
dtype: int64

### 3. Dealing With Categorical

In [36]:
cat = loan_export_copy.select_dtypes(include=['object'])

In [37]:
cat

FirstTimeHomebuyer    MSA Occupancy Channel PropertyType PostalCode  \
0                       N  16974         O       T           SF      60400   
1                       N  19740         O       R           SF      80200   
2                       N  29940         O       T           SF      66000   
3                       N  31084         O       R           SF      90700   
4                       N  35644         O       R           SF       7600   
...                   ...    ...       ...     ...          ...        ...   
291446                  N  37964         O       R           SF      19400   
291447                  N  35644         O       C           SF      11600   
291448                  N  42044         O       R           SF      92800   
291449                  N  37964         I       R           SF      19300   
291450                  N  37964         O       R           SF      18900   

       LoanPurpose NumBorrowers  
0                P            2  
1                N            1  
2                N            2  
3                N            2  
4                N            2  
...            ...          ...  
291446           N            2  
291447           P            2  
291448           N            1  
291449           P            2  
291450           P            2  

[291445 rows x 8 columns]

In [38]:
to_integer = 'MSA', 'PostalCode', 'NumBorrowers'

In [39]:
def convert_columns_to_integer(data_frame, columns):
    for column in columns:
        data_frame[column] = pd.to_numeric(data_frame[column], errors='coerce', downcast='integer')
    return data_frame

In [40]:
loan_export_copy = convert_columns_to_integer(loan_export_copy, to_integer)

In [41]:
loan_export_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291445 entries, 0 to 291450
Data columns (total 15 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   CreditScore         291445 non-null  int64  
 1   FirstTimeHomebuyer  291445 non-null  object 
 2   MSA                 291445 non-null  int32  
 3   Units               291445 non-null  int64  
 4   Occupancy           291445 non-null  object 
 5   OCLTV               291445 non-null  int64  
 6   DTI                 291445 non-null  int64  
 7   Channel             291445 non-null  object 
 8   PropertyType        291445 non-null  object 
 9   PostalCode          291445 non-null  int32  
 10  LoanPurpose         291445 non-null  object 
 11  NumBorrowers        291106 non-null  float64
 12  EverDelinquent      291445 non-null  int64  
 13  MonthsDelinquent    291445 non-null  int64  
 14  MonthsInRepayment   291445 non-null  int64  
dtypes: float64(1), int32(2), int64(7),

In [42]:
one_hot_encoding = ['FirstTimeHomebuyer', 'Occupancy', 'Channel', 'PropertyType', 'LoanPurpose', 'NumBorrowers']

In [43]:
def one_hot_encode_columns(data_frame, columns):
    encoded_df = pd.get_dummies(data_frame, columns=columns)
    return encoded_df

In [44]:
loan_export_copy = one_hot_encode_columns(loan_export_copy, one_hot_encoding)

In [45]:
loan_export_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291445 entries, 0 to 291450
Data columns (total 30 columns):
 #   Column                      Non-Null Count   Dtype
---  ------                      --------------   -----
 0   CreditScore                 291445 non-null  int64
 1   MSA                         291445 non-null  int32
 2   Units                       291445 non-null  int64
 3   OCLTV                       291445 non-null  int64
 4   DTI                         291445 non-null  int64
 5   PostalCode                  291445 non-null  int32
 6   EverDelinquent              291445 non-null  int64
 7   MonthsDelinquent            291445 non-null  int64
 8   MonthsInRepayment           291445 non-null  int64
 9   FirstTimeHomebuyer_N        291445 non-null  uint8
 10  FirstTimeHomebuyer_Y        291445 non-null  uint8
 11  Occupancy_I                 291445 non-null  uint8
 12  Occupancy_O                 291445 non-null  uint8
 13  Occupancy_S                 291445 non-null 

### 4. Preparing Data for Models

In [46]:
features, y = loan_export_copy.drop(['EverDelinquent'], axis=1), loan_export_copy.EverDelinquent.values

In [47]:
scaler = StandardScaler()
features = scaler.fit_transform(features)